synapse/handlers/federation.py

```diff
@@ -952,7 +952,20 @@ async def on_make_join_request(
         #
         # Note that this requires the /send_join request to come back to the
         # same server.
+        prev_event_ids = None
         if room_version.msc3083_join_rules:
+            # Note that the room's state can change out from under us and render our
+            # nice join rules-conformant event non-conformant by the time we build the
+            # event. When this happens, our validation at the end fails and we respond
+            # to the requesting server with a 403, which is misleading — it indicates
+            # that the user is not allowed to join the room and the joining server
+            # should not bother retrying via this homeserver or any others, when
+            # in fact we've just messed up with building the event.
+            #
+            # To reduce the likelihood of this race, we capture the forward extremities
+            # of the room (prev_event_ids) just before fetching the current state, and
```

Comment:
That's annoying, since the prev events we feed in won't always be representative of the room state the event is evaluated against.

Though I note that right now we hit an assert if there are more than 10 prev events.
https://github.com/matrix-org/synapse/blob/5febf88b6c5194582f427142dc0850625547c0d9/synapse/handlers/message.py#L1164-L1169

Reply:
Good to know. FWIW The spec says

> [Must contain less than or equal to 20 events.](https://spec.matrix.org/unstable/rooms/v10/#event-format-1)

Reply:
> That's annoying, since the prev events we feed in won't always be representative of the room state the event is evaluated against.

It turns out that `get_prev_events_for_room` only returns up to 10 forward extremities, so the prev events we feed in to the event builder will be the ones used. I hadn't quite realised that `get_prev_events_for_room` was subtly different to fetching the forward extremities.

https://github.com/matrix-org/synapse/blob/d0c713cc85f094c323b2ba3f02d8ac411a7f0705/synapse/storage/databases/main/event_federation.py#L1087-L1099